DirectX: Fixed multisample clamping logic

zigzag312 opened the pull request:
Bigger changes are that I moved some things from `PlatformInitialize()` to `PlatformSetup()` in GraphicsDevice.DirectX.cs. I hope I didn't move too much. Someone please verify as I'm new to this project. 

The reason for this change is that `d3ddevice `is needed before so we can query device capabilities. I compared DirectX version to OpenGL version. GL extension are set inside `PlatformSetup()` and GL Extensions are in some way like d3ddevice .... they are used to query device capabilities. So I tried to mimic OpenGL version. 

Multisample count clamping logic is removed from `GraphicsDeviceManager `and added it to `GraphicsDevice.DirectX` like @Jjagg said in #5449. It should also prevent crashes when multisampling is not supported like PR #5446.

KonajuGames commented on MonoGame.Framework/Graphics/GraphicsDevice.DirectX.cs:
```diff
@@ -130,24 +127,35 @@ private void PlatformInitialize()
+#endif
+#if WINDOWS
+            CreateSizeDependentResources();
+#endif
```
This can be cleaned up so it is a single

````
#if !WINDOWS_PHONE
        CreateSizeDependentResources();
#endif
````

zigzag312 commented:
Ok I will clean up those conditions and try to rebase

Jjagg commented on MonoGame.Framework/Graphics/GraphicsDevice.DirectX.cs:
```diff
@@ -827,13 +808,56 @@ internal void CreateSizeDependentResources()
         }
 
 #endif // WINDOWS
+        /// <returns>Higher than zero if multiSampleCount is supported. 
```
Comment is cut off here.

Jjagg commented on MonoGame.Framework/Graphics/GraphicsDevice.DirectX.cs:
```diff
@@ -827,13 +808,56 @@ internal void CreateSizeDependentResources()
+            var quality = _d3dDevice.CheckMultisampleQualityLevels(format, multiSampleCount) - 1;
+            // NOTE: should we always return highest quality?
+            return Math.Max(quality, 0); // clamp minimum to 0 
+        }
```
Newline here.

Jjagg commented on MonoGame.Framework/Graphics/GraphicsDevice.DirectX.cs:
```diff
@@ -827,13 +808,56 @@ internal void CreateSizeDependentResources()
+            }
+
+            return multisampleDesc;
+        }
```
Since the clamping is platform agnostic it could be put in the shared GraphicsDevice.cs file so other platforms can reuse it. (wihtou the SampleDescription obviously)

Jjagg commented on MonoGame.Framework/GraphicsDeviceManager.cs:
```diff
@@ -210,28 +210,6 @@ private GraphicsDeviceInformation DoPreparingDeviceSettings()
-                            msc = GraphicsDevice.GraphicsCapabilities.MaxMultiSampleCount;
-                    }
-                    gdi.PresentationParameters.MultiSampleCount = msc;
-                }
```
@zigzag312 I think you can just put this back in.

zigzag312 replied:
A have now added a call to clamping method inside Graphic.Reset instead. It is also called from ApplyChanges like DoPreparingDeviceSettings() and it also prevents wrong multi sample count if someone just calls Reset with new PresentationParameters.

Jjagg commented on MonoGame.Framework/Graphics/GraphicsCapabilities.DirectX.cs:
```diff
@@ -25,11 +25,13 @@ private void PlatformInitialize(GraphicsDevice device)
         private void PlatformInitializeAfterResources(GraphicsDevice device)
         {
-            _maxMultiSampleCount = GetMaxMultiSampleCount(device);
+            
```
Only DX platform used this method. If you can keep this change we can remove it. It was a bit of a hack anyway.

zigzag312 commented:
Ok I think I am getting a hang of this :)
I still need to change some things so it will behave like XNA.

KonajuGames commented:
> I still need to change some things so it will behave like XNA.

For sure.  Don't let the PR grow beyond scope though.  Keep it focused on the original intent and it will be easier to merge when it is complete.

zigzag312 commented:
After doing some testing I found out some differences between MG and XNA that are not connected just to multisampling. 

If we change some parameters of `PresentationParameters `under `GraphicsDevice `and call `GraphicsDevice.Reset()` XNA will ignore those changes, but MG will apply those changes. If we call `GraphicsDevice.Reset(newPresentationParameters)` both MG and XNA will apply `newPresentationParameters`. That is why a call to `ClampMultisampleCountToSupportedByCurrentDevice()` is needed right now inside Reset() method.

With this in mind, I think this PR should be merged as is (if there aren't other issues). And if `GraphicsDevice.Reset()` is modified to behave like in XNA in the future, then `ClampMultisampleCountToSupportedByCurrentDevice()` method will be moved out of `Reset()` method to some other location.

Jjagg commented on MonoGame.Framework/Graphics/GraphicsDevice.cs:
```diff
@@ -267,10 +267,34 @@ private void Setup()
+                // for very large numbers. That doesn't matter because
+                // the number will get clamped below anyway in this case.
+                // Also MultiSampleCount should never be larger than 32 
+                // (which is MultiSampleCountLimit)
```
The last two comment lines here aren't relevant anymore.

zigzag312 replied:
Removed

Jjagg commented:
This looks good to me. Would be nice to do some testing or add unit tests though.

> If we change some parameters of PresentationParametersunder GraphicsDeviceand call GraphicsDevice.Reset() XNA will ignore those changes, but MG will apply those changes.

Maybe the PresentationParameters getter makes a copy in XNA. IMO this isn't that big a problem.

zigzag312 commented:
Yeah, because XNA itself doesn't handle this particularly well. I actually prefer the way MG behaves in this case.

Any idea how to do unit tests for this?

KonajuGames commented:
Unit tests are difficult because it depends on GPU support for
multisampling. We use the DirectX reference device for unit tests, and I
believe it would have full support for multisampling.

Jjagg commented:
I meant for the proper clamping, so test if you can't mess up PresentationParameters somehow. But now that you mention it tests could break if the testing device does not support multisampling. Maybe just enable tests for DX with the reference device then.

zigzag312 commented:
Quick question. Should I split unit tests I just added? Because in each submitted test there are two (similar) tests. One for GraphicsDevice.ApplyChanges() and one for GraphicsDevice.Reset(PresentationParameters).

Jjagg commented on MonoGame.Framework/Graphics/GraphicsDevice.cs:
```diff
@@ -267,10 +267,39 @@ private void Setup()
+                GetClampedMultisampleCountToSupportedByCurrentDevice
+                    (PresentationParameters.MultiSampleCount);
+        }
+        internal int GetClampedMultisampleCountToSupportedByCurrentDevice(
```
@zigzag312  Can you make this private?

zigzag312 replied:
I need to check. I made some methods internal in the last refactoring so that same clamping logic is reused when creating multisampled render targets. I'm going to submit PR for that soon.

Jjagg commented on MonoGame.Framework/Graphics/GraphicsDevice.cs:
```diff
@@ -267,10 +267,39 @@ private void Setup()
             Dispose(false);
         }
 
+        void ClampMultisampleCountToSupportedByCurrentDevice()
```
I would inline this call. IMO it's not worth a seperate function and GraphicsDevice is cluttered enough as it is.

zigzag312 replied:
Ideally this method would be ClampPresentationParameters where we clamp all preferred values every time PresentationParameters are changed. Do we have any method like this?

Jjagg commented:
> Should I split unit tests I just added?

I think it's fine the way it is.

The only qualm I have with this PR in the current state is with the function name: `GetClampedMultisampleCountToSupportedByCurrentDevice` :p That's really longer than it needs to be IMO I vote for `ClampMultiSampleCount`.

Great work with this @zigzag312. Someone on the forums ran into the no suitable graphics device exception http://community.monogame.net/t/run-causes-no-suitable-graphics-device-exception-failed-to-create-graphics-device-3-6-0-1625-resolved/8819/2
Let's get this merged for 3.6.1!

KonajuGames commented:
Excessively wordy method names should be avoided. You can safely assume it
is for the current device.

zigzag312 commented:
Thanks! Yeah its a long name :) I'll shorten it.

zigzag312 commented:
I think this should be good to go now

Jjagg commented:
@KonajuGames @tomspilman Just a reminder that this should be in the 3.6.1 release. 
Without it DX/UWP crash when PreferMultisampling is set to true on the GraphicsDeviceManager and the GPU does not support a MS count of 32.

KonajuGames commented:
Merging this now.  Thanks @zigzag312
